File: 0065-valid-number/0065-valid-number.py
class Solution(object):
  def isNumber(self, s):
    """
    :type s: str
    :rtype: bool
    """
    digitsRe = re.compile("[0-9]")

    def trimSign(s):
      """
      :type s: str
      :rtype: str
      """
      if len(s) > 0 and s[0] in "+-":
        return s[1:]
      else:
        return s
    
    def validNum(s, numType):
      """
      :type s: str
      :type numType: str
      :rtype: bool
      """
      if len(s) == 0:
        return False
      
      if numType == "dec":
        dotIdx = -1
        for i in range(len(s)):
          if s[i] == ".":
            if dotIdx >= 0 or len(s) == 1:
              return False
            dotIdx = i
          elif re.match(digitsRe, s[i]) is None:
            return False
      elif numType == "int":
        for i in range(len(s)):
          if re.match(digitsRe, s[i]) is None:
            return False
      return True
    
    parts = [trimSign(part) for part in s.lower().split("e")]
    if len(parts) == 1:
      return validNum(parts[0], "dec")
    elif len(parts) == 2:
      return validNum(parts[0], "dec") and validNum(parts[1], "int")
    else:
      return False
      

File: 0065-valid-number/0065-valid-number.ts
function isNumber(s: string): boolean {
  function trimSign(s: string): string {
    if ("+-".includes(s[0])) {
      return s.slice(1)
    } else {
      return s
    }
  }

  function validNum(s: string, numType: string): boolean {
    if (s.length === 0) return false
    if (numType === "dec") {
      let dotIdx: number = -1
      for (let i = 0; i < s.length; i++) {
        if (s[i] === ".") {
          if (dotIdx >= 0 || s.length === 1) return false
          dotIdx = i
        } else if (!/[0-9]/.test(s[i])) {
          return false
        }
      }
    } else if (numType === "int") {
      for (let i = 0; i < s.length; i++) {
        if (!/[0-9]/.test(s[i])) return false
      }
    }
    return true
  }
  
  const parts: string[] = s.toLowerCase().split("e").map(part => trimSign(part))
  if (parts.length === 1) {
    return validNum(parts[0], "dec")
  } else if (parts.length === 2) {
    return validNum(parts[0], "dec") && validNum(parts[1], "int")
  } else {
    return false
  }
}